fennel/test_lib/data_engine.py

```diff
@@ -0,0 +1,607 @@
+                    f"Dataset {dataset._name} has no pipelines and is not a source dataset"
+                )
+            selected_pipelines = [
+                x
+                for x in dataset._pipelines
+                if x.tier.is_entity_selected(tier) and x.active
+            ]
+            sync_validation_for_pipelines(selected_pipelines, dataset._name)
+
+            for pipeline in selected_pipelines:
+                for input in pipeline.inputs:
+                    input_datasets_for_pipelines[input._name].append(
+                        f"{pipeline._dataset_name}.{pipeline.name}"
+                    )
+                    self.dataset_listeners[input._name].append(pipeline)
+
+        # Check that input_datasets_for_pipelines is a subset of self.datasets.
+        for ds, pipelines in input_datasets_for_pipelines.items():
+            if ds not in self.datasets:
+                raise ValueError(
+                    f"Dataset `{ds}` is an input to the pipelines: `{pipelines}` but is not synced. Please add it to the sync call."
+                )
+
+    def get_dataset_names(self) -> List[str]:
+        return list(self.datasets.keys())
+
+    def log(
+        self,
+        webhook: str,
+        endpoint: str,
+        df: pd.DataFrame,
+        _batch_size: int = 1000,
+    ) -> FakeResponse:
+        if df.shape[0] == 0:
+            print(f"Skipping log of empty dataframe for webhook {webhook}")
+            return FakeResponse(200, "OK")
+
+        webhook_endpoint = f"{webhook}:{endpoint}"
+        if webhook_endpoint not in self.webhook_to_dataset_map:
+            return FakeResponse(
+                404, f"Webhook endpoint {webhook_endpoint} not " f"found"
+            )
+        for ds in self.webhook_to_dataset_map[webhook_endpoint]:
+            try:
+                schema = self.datasets[ds].core_dataset.dsschema
+                if self.datasets[ds].pre_proc is not None:
+                    pre_proc_cols = list(
+                        self.datasets[ds].pre_proc.keys()  # type: ignore
+                    )
+                else:
+                    pre_proc_cols = []
+                df = cast_df_to_schema(df, schema, pre_proc_cols)
+            except Exception as e:
+                raise Exception(
+                    f"Schema validation failed during data insertion to `{ds}`: {str(e)}",
+                )
+            resp = self._internal_log(ds, df)
+            if resp.status_code != 200:
+                return resp
+        return FakeResponse(200, "OK")
+
+    def lookup(self, dataset_name: str, ts: pd.Series, *args, **kwargs):
+        if dataset_name not in self.datasets:
+            raise KeyError(f"Dataset: {dataset_name} not found")
+
+        fennel.datasets.datasets.dataset_lookup = partial(
+            self._dataset_lookup_impl,
+            [dataset_name],
+            None,
+        )
+
+        timestamps = pd.to_datetime(ts.apply(lambda x: parse_datetime(x)))
```

Comment:
can we use the cast_col_to_dtype function instead ?

Reply:
makes sense